### DISCUSSION TITLE
Image is locked up after converting to PDF

### DISCUSSION THREAD

**Discussion by tomasr78** [dlemstra/Magick.NET#1024] [Help]
When I try to convert Image to PDF sometimes the file is not released after conversion and I get an error that the file is used.

The code below demonstrates the issue and the second block of the code `File.Open` I am using to test if the file is in use.
I get error:

```
System.IO.IOExceptionMessageThe process cannot access the file 'C:\Temp\bXR1MJo1\image.pdf' because it is being used by another process.StackTraceString   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
```

Test code

```
            var magickReadSettings = new MagickReadSettings();
            const MagickFormat magickFormat = MagickFormat.Pdf;
            var destinationFileName = $@"{DestinationLocation}\image.{magickFormat.ToString().ToLower()}";

            using (var image = new MagickImage(Conversion.SourceFile.ToString(), magickReadSettings))
            {
                image.Settings.SetDefines(new PdfWriteDefines
                {
                    Author = "Image",
                    Producer = "Magick.net"
                });
                
                var colorSpace = ConverterParameter.ColorSpace();
                if (!Conversion.IsParameterSet(colorSpace))
                    image.ColorSpace = Conversion.ToString(colorSpace).ToEnum<ColorSpace>();
                image.Write(destinationFileName, magickFormat);
            }

        
         try
            {
                using(var stream = File.Open(destinationFileName, FileMode.Open, FileAccess.Read, FileShare.Read))
                {
                    stream.Close();
                }
            }
            catch (Exception e)
            {
                Logger.Main.LogError($"The file is locked FileMode.Open, FileAccess.Read, FileShare.Read {destinationFileName}",e);
            }
```

**Comment by dlemstra** [comment]
Is this reproducable with a specific input file?

**Reply by tomasr78** [reply]
No, this happens with random files and our app executes several code instances on different threds.  The strange thing is that when we log in to our server the file is accessible and it is not locked anymore. The bad thing is that I can't debug on the prod machine and can't reproduce on dev.

**Reply by tomasr78** [reply]
@dlemstra Is it possible to get ImageMagick PID from Magick.net while converting Image to PDF? I suppose you use ImageMagick under the hood for PDF conversion and I think that process is not completely released in some cases. I would try to find more information using PID and check is it still alive when an exception throw.

**Reply by dlemstra** [reply]
ImageMagick is executed through an API (interop call the C API) so it has no PID and will be executed in the same process. I did just find a spot where we did not close the file properly and this will be fixed in the next release. Not sure if that is your issue though.

**Reply by tomasr78** [reply]
When are you planning to release the next version? I will test it and let you know.

**Reply by dlemstra** [reply]
The new release has been published earlier this week.

**Reply by tomasr78** [reply]
Could you drop the link to that commit? I would like to review the fix.

**Reply by dlemstra** [reply]
This is the commit: https://github.com/ImageMagick/ImageMagick/commit/225b51d7f2e86b8c8588022fe1e391249d023904. And thanks for helping out with this.

**Reply by dlemstra** [reply]
Are you still experiencing this issue?

**Comment by tomasr78** [comment]
Sorry for the delay, I have tried to narrow down the problems and that's what I found.
It seems the problem is not related directly to Magick.net but to `Process.Start`. I suppose you also use `Process.Start` to launch ImageMagick.

So the problem is that Magick.Net and my console app use `Process.Start` at the same time to start other console apps and somehow they interact and lock each other files.

This problem only happens when `Process.StartInfo.UseShellExecute` is set to `False`. I already spend five days trying to find out what is causing the problem and can't find it yet.
```
    using (var p = new Process
    {
        StartInfo =
        {
            FileName = "c:\commandLine.exe",
            Arguments = "c:\file.jpg",
            UseShellExecute = false,
            RedirectStandardError = true,
            RedirectStandardOutput = true
        }
    })
    {
        try
        {
            p.Start();
            p.PriorityClass = ProcessPriorityClass.Normal;

            p.WaitForExit();

            var output = await p.StandardOutput.ReadToEndAsync();
            var error = await p.StandardError.ReadToEndAsync();
            var exitCode = p.ExitCode;
            return .....; 

        }
        finally
        {
            try
            {
                if (!p.HasExited)
                    p.Kill();
            }
            catch (InvalidOperationException)
            {
                //The process is not found, probably executable was not fount                    
            }

            p.Close();
        }
    } 
```

**Comment by tomasr78** [comment]
Dirk, could you show me the code where you run ImageMagick? I think there is a bug because if we remove Magick.Net our Processes start working fine and files are not locked anymore.
